lib/Model/Importer/DebugInfo/DwarfImporter.cpp

```diff
@@ -451,83 +421,78 @@ class DwarfToModelConverter : public BinaryImporterHelper {
     return false;
   }
 
-  RecursiveCoroutine<const model::QualifiedType *>
-  getType(const DWARFDie &Die) {
-    if (auto MaybeType = Die.find(DW_AT_type)) {
-      if (MaybeType->getForm() == llvm::dwarf::DW_FORM_GNU_ref_alt) {
-        rc_return findAltType(MaybeType->getRawUValue());
+  RecursiveCoroutine<model::UpcastableType> makeType(const DWARFDie &Die) {
+    if (auto Type = Die.find(DW_AT_type)) {
+      if (Type->getForm() == llvm::dwarf::DW_FORM_GNU_ref_alt) {
+        rc_return Importer.findType({ AltIndex, Type->getRawUValue() }).copy();
       } else {
-        DWARFDie InnerDie = DICtx.getDIEForOffset(*MaybeType->getAsReference());
+        DWARFDie InnerDie = DICtx.getDIEForOffset(*Type->getAsReference());
         rc_return rc_recur resolveType(InnerDie, false);
       }
     } else if (auto MaybeOrigin = Die.find(DW_AT_abstract_origin)) {
       DWARFDie Origin = DICtx.getDIEForOffset(*MaybeOrigin->getAsReference());
-      rc_return rc_recur getType(Origin);
+      rc_return rc_recur makeType(Origin);
     } else {
       rc_return nullptr;
     }
   }
 
-  RecursiveCoroutine<model::QualifiedType> getTypeOrVoid(const DWARFDie &Die) {
-    using namespace model;
-    using PTK = model::PrimitiveKind::Values;
-    const model::QualifiedType *Result = rc_recur getType(Die);
-    if (Result != nullptr) {
-      rc_return *Result;
-    } else {
-      rc_return QualifiedType(Model->getPrimitiveType(PTK::Void, 0), {});
-    }
+  RecursiveCoroutine<model::UpcastableType>
+  makeTypeOrVoid(const DWARFDie &Die) {
+    if (model::UpcastableType Result = rc_recur makeType(Die))
+      rc_return Result;
+    else
+      rc_return model::PrimitiveType::makeVoid();
   }
 
-  RecursiveCoroutine<const model::QualifiedType *>
-  resolveTypeWithIdentity(const DWARFDie &Die,
-                          model::QualifiedType *DefinitionReference) {
+  RecursiveCoroutine<bool>
+  resolveTypeWithIdentity(const DWARFDie &Die, model::UpcastableType &Type) {
     using namespace model;
 
     auto Offset = Die.getOffset();
     auto Tag = Die.getTag();
 
     revng_assert(Placeholders.contains(Offset));
-    revng_assert(DefinitionReference->Qualifiers().empty());
-    model::TypeDefinition *T = DefinitionReference->UnqualifiedType().get();
+    model::TypeDefinition *Definition = Type->tryGetAsDefinition();
+    revng_assert(Definition != nullptr);
```

Comment:
Can't we use `asDefinition` here?

Reply:
Ah, right, I converted this back because I was debugging this bit.

Thanks for noticing.

Reply:
Done.